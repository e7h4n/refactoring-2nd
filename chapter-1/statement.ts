interface Play {
    readonly name: string;
    readonly type: 'tragedy' | 'comedy';
}

interface Performance {
    readonly playID: string;
    readonly audience: number;
}

interface Invoice {
    readonly customer: string;

    readonly performances: Performance[];
}

interface PerformanceReport {
    readonly playName: string;

    readonly amount: number;

    readonly audience: number;

    readonly credits: number;
}

interface CustomerStatement {
    readonly customer: string;

    readonly amount: number;

    readonly credits: number;

    readonly performances: PerformanceReport[];
}

export function statement(invoice: Invoice, plays:  {[key: string]: Play}) {
    const customerStatement = genStatement(invoice, plays);

    return genTextReport(customerStatement);
}

function getAmount(playType: 'tragedy' | 'comedy', audience: number) {
    let amount = 0;

    switch (playType) {
        case 'tragedy':
            amount = 40000;

            if (audience > 30) {
                amount += 1000 * (audience - 30);
            }
        break;

        case 'comedy':
            amount = 30000;

            if (audience > 20) {
                amount += 10000 + 500 * (audience - 20);
            }

            amount += 300 * audience;
        break;
        default:
            throw new Error(`unknown type: ${playType}`);
    }

    return amount;
}

function getCredit(playType: 'tragedy' | 'comedy', audience: number) {
    let credits = Math.max(audience - 30, 0);
    // add extra credit for every then comedy attendees
    if ('comedy' === playType) {
        credits = credits + Math.floor(audience / 5);
    }

    return credits;
}

function getTotalCredit(reports: PerformanceReport[]) {
    let totalCredits = 0;
    for (let perfReport of reports) {
        totalCredits = totalCredits + perfReport.credits;
    }
    return totalCredits;
}

function getTotalAmount(reports: PerformanceReport[]) {
    let totalAmount = 0;
    for (let perfReport of reports) {
        totalAmount = totalAmount + perfReport.amount;
    }
    return totalAmount;
}

function getPerformanceReports(performances: Performance[], plays: {[key: string]: Play}) {
    return performances.map(perf => {
        const play = plays[perf.playID];
        return getPerformanceReport(perf, play);
    });
}

function getPerformanceReport(perf: Performance, play: Play) {
    const credits = getCredit(play.type, perf.audience);
    const amount = getAmount(play.type, perf.audience);

    return {
        playName: play.name,
        amount: amount,
        credits: credits,
        audience: perf.audience,
    };
}

function currency(num: number) {
    const format = new Intl.NumberFormat('en-US', {
        style: 'currency',
        currency: 'USD',
        minimumFractionDigits: 2
    }).format;

    return format(num/100);
}

function genStatement(invoice: Invoice, plays: {[key: string]: Play}): CustomerStatement {
    const perfReports: PerformanceReport[] = getPerformanceReports(invoice.performances, plays);

    return {
        customer: invoice.customer,
        amount: getTotalAmount(perfReports),
        credits: getTotalCredit(perfReports),
        performances: perfReports,
    };
}

function genTextReport(customerStatement: CustomerStatement) {
    let result = `Statement for ${customerStatement.customer}\n`;
    for (const perfReport of customerStatement.performances) {
        result += ` ${perfReport.playName}: ${currency(perfReport.amount)} (${perfReport.audience} seats)\n`;
    }

    result += `Amount owed is ${currency(customerStatement.amount)}\n`;
    result += `You earned ${customerStatement.credits} credits\n`;
    return result;
}
