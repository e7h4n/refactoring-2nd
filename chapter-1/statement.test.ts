import { statement } from './statement';

function verifyAllCombinationsSnapshot(testName: string, targetFunc, ...argsList: any[][]) {
    const result: any[][] = [];
    const max = argsList.length - 1;

    function helper(arr: any[][], i: number) {
        for (let j = 0, l = argsList[i].length; j < l; j++) {
            const a = arr.slice(0);
            a.push(argsList[i][j]);
            if (i === max)
                result.push(a);
            else
                helper(a, i+1);
        }
    }

    helper([], 0);

    for(let i = 0; i < result.length; i++) {
        test(testName + ': ' + JSON.stringify(result[i]), async () => {
            let ret = null;
            try {
                ret = await targetFunc(...result[i]);
                expect(ret).toMatchSnapshot();
            } catch (e) {
                expect(e).toMatchSnapshot();
            }
        });
    }
}


describe('statement', () => {
    function performanceFooWithAudience(audience: number | null) {
        if (audience === null) {
            return {
                performances: [],
            };
        }

        return {
            performances: [{
                playID: 'foo',
                audience,
            }],
        }
    }

    function playFooWithType(playType: 'comedy' | 'tragedy') {
        return {
            foo: {
                name: 'Foo',
                type: playType,
            }
        }
    }

    verifyAllCombinationsSnapshot('statement safety net', (audience: number | null, type: string | null) => {
        const perfs = performanceFooWithAudience(audience)
        const plays = playFooWithType(type);
        return statement(perfs, plays);
    }, [
        null,
        10,
        100,
        20,
        21,
        29,
        31,
        30,
    ], [
        null,
        'comedy',
        'tragedy',
        'foo',
    ]);
});
