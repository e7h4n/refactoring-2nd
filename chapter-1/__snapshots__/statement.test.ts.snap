// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`statement statement safety net: [10,"comedy"] 1`] = `
"Statement for undefined
 Foo: $330.00 (10 seats)
Amount owed is $330.00
You earned 2 credits
"
`;

exports[`statement statement safety net: [10,"foo"] 1`] = `[Error: unknown type: foo]`;

exports[`statement statement safety net: [10,"tragedy"] 1`] = `
"Statement for undefined
 Foo: $400.00 (10 seats)
Amount owed is $400.00
You earned 0 credits
"
`;

exports[`statement statement safety net: [10,null] 1`] = `[Error: unknown type: null]`;

exports[`statement statement safety net: [20,"comedy"] 1`] = `
"Statement for undefined
 Foo: $360.00 (20 seats)
Amount owed is $360.00
You earned 4 credits
"
`;

exports[`statement statement safety net: [20,"foo"] 1`] = `[Error: unknown type: foo]`;

exports[`statement statement safety net: [20,"tragedy"] 1`] = `
"Statement for undefined
 Foo: $400.00 (20 seats)
Amount owed is $400.00
You earned 0 credits
"
`;

exports[`statement statement safety net: [20,null] 1`] = `[Error: unknown type: null]`;

exports[`statement statement safety net: [21,"comedy"] 1`] = `
"Statement for undefined
 Foo: $468.00 (21 seats)
Amount owed is $468.00
You earned 4 credits
"
`;

exports[`statement statement safety net: [21,"foo"] 1`] = `[Error: unknown type: foo]`;

exports[`statement statement safety net: [21,"tragedy"] 1`] = `
"Statement for undefined
 Foo: $400.00 (21 seats)
Amount owed is $400.00
You earned 0 credits
"
`;

exports[`statement statement safety net: [21,null] 1`] = `[Error: unknown type: null]`;

exports[`statement statement safety net: [29,"comedy"] 1`] = `
"Statement for undefined
 Foo: $532.00 (29 seats)
Amount owed is $532.00
You earned 5 credits
"
`;

exports[`statement statement safety net: [29,"foo"] 1`] = `[Error: unknown type: foo]`;

exports[`statement statement safety net: [29,"tragedy"] 1`] = `
"Statement for undefined
 Foo: $400.00 (29 seats)
Amount owed is $400.00
You earned 0 credits
"
`;

exports[`statement statement safety net: [29,null] 1`] = `[Error: unknown type: null]`;

exports[`statement statement safety net: [30,"comedy"] 1`] = `
"Statement for undefined
 Foo: $540.00 (30 seats)
Amount owed is $540.00
You earned 6 credits
"
`;

exports[`statement statement safety net: [30,"foo"] 1`] = `[Error: unknown type: foo]`;

exports[`statement statement safety net: [30,"tragedy"] 1`] = `
"Statement for undefined
 Foo: $400.00 (30 seats)
Amount owed is $400.00
You earned 0 credits
"
`;

exports[`statement statement safety net: [30,null] 1`] = `[Error: unknown type: null]`;

exports[`statement statement safety net: [31,"comedy"] 1`] = `
"Statement for undefined
 Foo: $548.00 (31 seats)
Amount owed is $548.00
You earned 7 credits
"
`;

exports[`statement statement safety net: [31,"foo"] 1`] = `[Error: unknown type: foo]`;

exports[`statement statement safety net: [31,"tragedy"] 1`] = `
"Statement for undefined
 Foo: $410.00 (31 seats)
Amount owed is $410.00
You earned 1 credits
"
`;

exports[`statement statement safety net: [31,null] 1`] = `[Error: unknown type: null]`;

exports[`statement statement safety net: [100,"comedy"] 1`] = `
"Statement for undefined
 Foo: $1,100.00 (100 seats)
Amount owed is $1,100.00
You earned 90 credits
"
`;

exports[`statement statement safety net: [100,"foo"] 1`] = `[Error: unknown type: foo]`;

exports[`statement statement safety net: [100,"tragedy"] 1`] = `
"Statement for undefined
 Foo: $1,100.00 (100 seats)
Amount owed is $1,100.00
You earned 70 credits
"
`;

exports[`statement statement safety net: [100,null] 1`] = `[Error: unknown type: null]`;

exports[`statement statement safety net: [null,"comedy"] 1`] = `
"Statement for undefined
Amount owed is $0.00
You earned 0 credits
"
`;

exports[`statement statement safety net: [null,"foo"] 1`] = `
"Statement for undefined
Amount owed is $0.00
You earned 0 credits
"
`;

exports[`statement statement safety net: [null,"tragedy"] 1`] = `
"Statement for undefined
Amount owed is $0.00
You earned 0 credits
"
`;

exports[`statement statement safety net: [null,null] 1`] = `
"Statement for undefined
Amount owed is $0.00
You earned 0 credits
"
`;
